// @wordpress/dom/build-module/dom/index.mjs
// packages/dom/src/dom/index.js
import { default as default2 } from "./compute-caret-rect.mjs";
import { default as default3 } from "./document-has-text-selection.mjs";
import { default as default4 } from "./document-has-uncollapsed-selection.mjs";
import { default as default5 } from "./document-has-selection.mjs";
import { default as default6 } from "./get-rectangle-from-range.mjs";
import { default as default7 } from "./get-scroll-container.mjs";
import { default as default8 } from "./get-offset-parent.mjs";
import { default as default9 } from "./is-entirely-selected.mjs";
import { default as default10 } from "./is-form-element.mjs";
import { default as default11 } from "./is-horizontal-edge.mjs";
import { default as default12 } from "./is-number-input.mjs";
import { default as default13 } from "./is-text-field.mjs";
import { default as default14 } from "./is-vertical-edge.mjs";
import { default as default15 } from "./place-caret-at-horizontal-edge.mjs";
import { default as default16 } from "./place-caret-at-vertical-edge.mjs";
import { default as default17 } from "./replace.mjs";
import { default as default18 } from "./remove.mjs";
import { default as default19 } from "./insert-after.mjs";
import { default as default20 } from "./unwrap.mjs";
import { default as default21 } from "./replace-tag.mjs";
import { default as default22 } from "./wrap.mjs";
import { default as default23 } from "./strip-html.mjs";
import { default as default24 } from "./is-empty.mjs";
import { default as default25 } from "./remove-invalid-html.mjs";
import { default as default26 } from "./is-rtl.mjs";
import { default as default27 } from "./safe-html.mjs";
import { default as default28 } from "./is-selection-forward.mjs";
export {
  default23 as __unstableStripHTML,
  default2 as computeCaretRect,
  default5 as documentHasSelection,
  default3 as documentHasTextSelection,
  default4 as documentHasUncollapsedSelection,
  default8 as getOffsetParent,
  default6 as getRectangleFromRange,
  default7 as getScrollContainer,
  default19 as insertAfter,
  default24 as isEmpty,
  default9 as isEntirelySelected,
  default10 as isFormElement,
  default11 as isHorizontalEdge,
  default12 as isNumberInput,
  default26 as isRTL,
  default28 as isSelectionForward,
  default13 as isTextField,
  default14 as isVerticalEdge,
  default15 as placeCaretAtHorizontalEdge,
  default16 as placeCaretAtVerticalEdge,
  default18 as remove,
  default25 as removeInvalidHTML,
  default17 as replace,
  default21 as replaceTag,
  default27 as safeHTML,
  default20 as unwrap,
  default22 as wrap
};
