// @wordpress/theme/build-module/theme-provider.mjs
// packages/theme/src/theme-provider.tsx
import { useMemo, useRef } from "@wordpress/element";
import { useIsomorphicLayoutEffect } from "@wordpress/compose";
import { ThemeContext } from "./context.mjs";
import { useThemeProviderStyles } from "./use-theme-provider-styles.mjs";

// packages/style-runtime/src/index.ts
var STYLE_HASH_ATTRIBUTE = "data-wp-hash";
function getRuntime() {
  const globalScope = globalThis;
  if (globalScope.__wpStyleRuntime) {
    return globalScope.__wpStyleRuntime;
  }
  globalScope.__wpStyleRuntime = {
    documents: /* @__PURE__ */ new Map(),
    styles: /* @__PURE__ */ new Map(),
    injectedStyles: /* @__PURE__ */ new WeakMap()
  };
  if (typeof document !== "undefined") {
    registerDocument(document);
  }
  return globalScope.__wpStyleRuntime;
}
function documentContainsStyleHash(targetDocument, hash) {
  if (!targetDocument.head) {
    return false;
  }
  for (const style of targetDocument.head.querySelectorAll(
    `style[${STYLE_HASH_ATTRIBUTE}]`
  )) {
    if (style.getAttribute(STYLE_HASH_ATTRIBUTE) === hash) {
      return true;
    }
  }
  return false;
}
function injectStyle(targetDocument, hash, css) {
  if (!targetDocument.head) {
    return;
  }
  const runtime = getRuntime();
  let injectedStyles = runtime.injectedStyles.get(targetDocument);
  if (!injectedStyles) {
    injectedStyles = /* @__PURE__ */ new Set();
    runtime.injectedStyles.set(targetDocument, injectedStyles);
  }
  if (injectedStyles.has(hash)) {
    return;
  }
  if (documentContainsStyleHash(targetDocument, hash)) {
    injectedStyles.add(hash);
    return;
  }
  const style = targetDocument.createElement("style");
  style.setAttribute(STYLE_HASH_ATTRIBUTE, hash);
  style.appendChild(targetDocument.createTextNode(css));
  targetDocument.head.appendChild(style);
  injectedStyles.add(hash);
}
function registerDocument(targetDocument) {
  const runtime = getRuntime();
  runtime.documents.set(
    targetDocument,
    (runtime.documents.get(targetDocument) ?? 0) + 1
  );
  for (const [hash, css] of runtime.styles) {
    injectStyle(targetDocument, hash, css);
  }
  return () => {
    const count = runtime.documents.get(targetDocument);
    if (count === void 0) {
      return;
    }
    if (count <= 1) {
      runtime.documents.delete(targetDocument);
      return;
    }
    runtime.documents.set(targetDocument, count - 1);
  };
}
function registerStyle(hash, css) {
  const runtime = getRuntime();
  runtime.styles.set(hash, css);
  for (const targetDocument of runtime.documents.keys()) {
    injectStyle(targetDocument, hash, css);
  }
}

// packages/theme/src/style.module.css
if (typeof process === "undefined" || process.env.NODE_ENV !== "test") {
  registerStyle("f4e6e06c6a", ".dba930ea7a9438fd__root{display:contents}");
}
var style_default = { "root": "dba930ea7a9438fd__root" };

// packages/theme/src/theme-provider.tsx
import { jsx } from "react/jsx-runtime";
var rootProviderCountByDocument = /* @__PURE__ */ new WeakMap();
var ThemeProvider = ({
  children,
  color = {},
  cursor,
  cornerRadius,
  isRoot = false
}) => {
  const { themeProviderStyles, resolvedSettings } = useThemeProviderStyles({
    color,
    cursor,
    cornerRadius
  });
  const cornerRadiusPreset = resolvedSettings.cornerRadius ?? "subtle";
  const contextValue = useMemo(
    () => ({
      resolvedSettings
    }),
    [resolvedSettings]
  );
  const wrapperRef = useRef(null);
  useIsomorphicLayoutEffect(() => {
    if (!isRoot) {
      return;
    }
    const doc = wrapperRef.current?.ownerDocument;
    if (!doc) {
      return;
    }
    const root = doc.documentElement;
    if (process.env.NODE_ENV !== "production") {
      const active = rootProviderCountByDocument.get(doc) ?? 0;
      if (active > 0) {
        console.warn(
          "ThemeProvider: More than one root provider (`isRoot`) is mounted on the same document. Their forwarded document-level styles conflict, and unmounting one can reset the others. Render a single root provider per document."
        );
      }
      rootProviderCountByDocument.set(doc, active + 1);
    }
    const previous = /* @__PURE__ */ new Map();
    const applied = [];
    for (const [rawKey, rawValue] of Object.entries(
      themeProviderStyles
    )) {
      if (!rawKey.startsWith("--") || rawValue === null || rawValue === void 0) {
        continue;
      }
      previous.set(rawKey, root.style.getPropertyValue(rawKey));
      root.style.setProperty(rawKey, String(rawValue));
      applied.push(rawKey);
    }
    return () => {
      if (process.env.NODE_ENV !== "production") {
        const active = rootProviderCountByDocument.get(doc) ?? 1;
        if (active <= 1) {
          rootProviderCountByDocument.delete(doc);
        } else {
          rootProviderCountByDocument.set(doc, active - 1);
        }
      }
      for (const key of applied) {
        const prev = previous.get(key);
        if (prev) {
          root.style.setProperty(key, prev);
        } else {
          root.style.removeProperty(key);
        }
      }
    };
  }, [isRoot, themeProviderStyles]);
  return /* @__PURE__ */ jsx(
    "div",
    {
      ref: wrapperRef,
      "data-wpds-root-provider": isRoot || void 0,
      "data-wpds-corner-radius": cornerRadiusPreset,
      className: style_default.root,
      style: themeProviderStyles,
      children: /* @__PURE__ */ jsx(ThemeContext.Provider, { value: contextValue, children })
    }
  );
};
export {
  ThemeProvider
};
