// @wordpress/theme/build-module/use-theme-provider-styles.mjs
// packages/theme/src/use-theme-provider-styles.ts
import {
  ColorSpace,
  clone,
  set,
  to,
  sRGB,
  HSL
} from "colorjs.io/fn";
import memoize from "memize";
import { useMemo, useContext } from "@wordpress/element";
import { ThemeContext } from "./context.mjs";
import colorTokens from "./prebuilt/ts/color-tokens.mjs";
import {
  buildBgRamp,
  buildAccentRamp,
  DEFAULT_SEED_COLORS
} from "./color-ramps/index.mjs";
import { getColorString } from "./color-ramps/lib/color-utils.mjs";
var getCachedBgRamp = memoize(buildBgRamp, { maxSize: 10 });
var getCachedAccentRamp = memoize(buildAccentRamp, { maxSize: 10 });
var legacyWpComponentsOverridesCSS = [
  ["--wp-components-color-accent", "var(--wp-admin-theme-color)"],
  [
    "--wp-components-color-accent-darker-10",
    "var(--wp-admin-theme-color-darker-10)"
  ],
  [
    "--wp-components-color-accent-darker-20",
    "var(--wp-admin-theme-color-darker-20)"
  ],
  [
    "--wp-components-color-accent-inverted",
    "var(--wpds-color-foreground-interactive-brand-strong, #fff)"
  ],
  [
    "--wp-components-color-background",
    "var(--wpds-color-background-surface-neutral-strong, #fff)"
  ],
  [
    "--wp-components-color-foreground",
    "var(--wpds-color-foreground-content-neutral, #1e1e1e)"
  ],
  [
    "--wp-components-color-foreground-inverted",
    "var(--wpds-color-background-surface-neutral, #fcfcfc)"
  ],
  [
    "--wp-components-color-gray-100",
    "var(--wpds-color-background-surface-neutral, #fcfcfc)"
  ],
  [
    "--wp-components-color-gray-200",
    "var(--wpds-color-stroke-surface-neutral, #dbdbdb)"
  ],
  [
    "--wp-components-color-gray-300",
    "var(--wpds-color-stroke-surface-neutral, #dbdbdb)"
  ],
  [
    "--wp-components-color-gray-400",
    "var(--wpds-color-stroke-interactive-neutral, #8d8d8d)"
  ],
  [
    "--wp-components-color-gray-600",
    "var(--wpds-color-stroke-interactive-neutral, #8d8d8d)"
  ],
  [
    "--wp-components-color-gray-700",
    "var(--wpds-color-foreground-content-neutral-weak, #707070)"
  ],
  [
    "--wp-components-color-gray-800",
    "var(--wpds-color-foreground-content-neutral, #1e1e1e)"
  ]
];
function customRgbFormat(color) {
  const rgb = to(color, sRGB);
  return rgb.coords.map((n) => Math.round((n ?? 0) * 255)).join(", ");
}
function legacyWpAdminThemeOverridesCSS(accent) {
  ColorSpace.register(sRGB);
  const parsedAccent = to(accent, HSL);
  const parsedL = parsedAccent.coords[2] ?? 0;
  const darker10 = set(
    clone(parsedAccent),
    [HSL, "l"],
    Math.max(0, parsedL - 5)
    // L reduced by 5%
  );
  const darker20 = set(
    clone(parsedAccent),
    [HSL, "l"],
    Math.max(0, parsedL - 10)
    // L reduced by 10%
  );
  return [
    ["--wp-admin-theme-color", getColorString(parsedAccent)],
    ["--wp-admin-theme-color--rgb", customRgbFormat(parsedAccent)],
    ["--wp-admin-theme-color-darker-10", getColorString(darker10)],
    [
      "--wp-admin-theme-color-darker-10--rgb",
      customRgbFormat(darker10)
    ],
    ["--wp-admin-theme-color-darker-20", getColorString(darker20)],
    [
      "--wp-admin-theme-color-darker-20--rgb",
      customRgbFormat(darker20)
    ]
  ];
}
function colorTokensCSS(computedColorRamps) {
  const entries = [];
  for (const [rampName, { ramp }] of computedColorRamps) {
    for (const [tokenName, tokenValue] of Object.entries(ramp)) {
      const key = `${rampName}-${tokenName}`;
      const aliasedBy = colorTokens[key] ?? [];
      for (const aliasedId of aliasedBy) {
        entries.push([`--wpds-color-${aliasedId}`, tokenValue]);
      }
    }
  }
  return entries;
}
function generateStyles({
  primary,
  computedColorRamps
}) {
  return Object.fromEntries(
    [
      // Semantic color tokens
      colorTokensCSS(computedColorRamps),
      // Legacy overrides
      legacyWpAdminThemeOverridesCSS(primary),
      legacyWpComponentsOverridesCSS
    ].flat()
  );
}
function useThemeProviderStyles({
  color = {},
  cursor,
  cornerRadius
} = {}) {
  const { resolvedSettings: inheritedSettings } = useContext(ThemeContext);
  const primary = color.primary ?? inheritedSettings.color?.primary ?? DEFAULT_SEED_COLORS.primary;
  const background = color.background ?? inheritedSettings.color?.background ?? DEFAULT_SEED_COLORS.background;
  const cursorControl = cursor?.control ?? inheritedSettings.cursor?.control;
  const cornerRadiusPreset = cornerRadius ?? inheritedSettings.cornerRadius ?? "subtle";
  const resolvedSettings = useMemo(
    () => ({
      color: {
        primary,
        background
      },
      cursor: cursorControl ? { control: cursorControl } : void 0,
      cornerRadius: cornerRadiusPreset
    }),
    [primary, background, cursorControl, cornerRadiusPreset]
  );
  const colorStyles = useMemo(() => {
    const seeds = {
      ...DEFAULT_SEED_COLORS,
      background,
      primary
    };
    const computedColorRamps = /* @__PURE__ */ new Map();
    const bgRamp = getCachedBgRamp(seeds.background);
    Object.entries(seeds).forEach(([rampName, seed]) => {
      if (rampName === "background") {
        computedColorRamps.set("bg", bgRamp);
      } else {
        computedColorRamps.set(
          rampName,
          getCachedAccentRamp(seed, bgRamp)
        );
      }
    });
    return generateStyles({
      primary: seeds.primary,
      computedColorRamps
    });
  }, [primary, background]);
  const themeProviderStyles = useMemo(
    () => ({
      ...colorStyles,
      ...cursorControl && {
        "--wpds-cursor-control": cursorControl
      }
    }),
    [colorStyles, cursorControl]
  );
  return {
    resolvedSettings,
    themeProviderStyles
  };
}
export {
  useThemeProviderStyles
};
