// @wordpress/e2e-test-utils-playwright/build/request-utils/preferences.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resetPreferences = resetPreferences;
exports.setPreferences = setPreferences;
/**
 * Reset user preferences
 *
 * @param this Request utils.
 */
async function resetPreferences() {
    await this.rest({
        path: '/wp/v2/users/me',
        method: 'PUT',
        data: {
            meta: {
                persisted_preferences: {},
            },
        },
    });
}
/**
 * Set scoped user preferences via the REST API.
 *
 * Reads the current persisted preferences, merges the new values under the
 * given scope, and writes them back.
 *
 * @param this        Request utils.
 * @param scope       The preference scope (e.g. 'core/edit-post').
 * @param preferences Key/value map of preference names to values.
 */
async function setPreferences(scope, preferences) {
    const user = await this.rest({
        path: '/wp/v2/users/me',
        method: 'GET',
        params: { context: 'edit' },
    });
    const currentPreferences = user.meta?.persisted_preferences ?? {};
    await this.rest({
        path: '/wp/v2/users/me',
        method: 'PUT',
        data: {
            meta: {
                persisted_preferences: {
                    ...currentPreferences,
                    [scope]: {
                        ...(currentPreferences[scope] ?? {}),
                        ...preferences,
                    },
                },
            },
        },
    });
}
