// @wordpress/compose/build/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/compose/src/index.js
var index_exports = {};
__export(index_exports, {
  __experimentalUseDialog: () => import_use_dialog.default,
  __experimentalUseDragging: () => import_use_dragging.default,
  __experimentalUseDropZone: () => import_use_drop_zone.default,
  __experimentalUseFixedWindowList: () => import_use_fixed_window_list.default,
  __experimentalUseFocusOutside: () => import_use_focus_outside.default,
  compose: () => import_compose.default,
  ifCondition: () => import_if_condition.default,
  pipe: () => import_pipe.default,
  privateApis: () => import_private_apis.privateApis,
  pure: () => import_pure.default,
  useAsyncList: () => import_use_async_list.default,
  useConstrainedTabbing: () => import_use_constrained_tabbing.default,
  useCopyOnClick: () => import_use_copy_on_click.default,
  useCopyToClipboard: () => import_use_copy_to_clipboard.default,
  useDebounce: () => import_use_debounce.default,
  useDebouncedInput: () => import_use_debounced_input.default,
  useDisabled: () => import_use_disabled.default,
  useEvent: () => import_use_event.default,
  useFocusOnMount: () => import_use_focus_on_mount.useFocusOnMount,
  useFocusReturn: () => import_use_focus_return.default,
  useFocusableIframe: () => import_use_focusable_iframe.default,
  useInstanceId: () => import_use_instance_id.default,
  useIsomorphicLayoutEffect: () => import_use_isomorphic_layout_effect.default,
  useKeyboardShortcut: () => import_use_keyboard_shortcut.default,
  useMediaQuery: () => import_use_media_query.default,
  useMergeRefs: () => import_use_merge_refs.default,
  useObservableValue: () => import_use_observable_value.default,
  usePrevious: () => import_use_previous.default,
  useReducedMotion: () => import_use_reduced_motion.default,
  useRefEffect: () => import_use_ref_effect.default,
  useResizeObserver: () => import_use_resize_observer.default,
  useStateWithHistory: () => import_use_state_with_history.default,
  useThrottle: () => import_use_throttle.default,
  useViewportMatch: () => import_use_viewport_match.default,
  useWarnOnChange: () => import_use_warn_on_change.default,
  withGlobalEvents: () => import_with_global_events.default,
  withInstanceId: () => import_with_instance_id.default,
  withSafeTimeout: () => import_with_safe_timeout.default,
  withState: () => import_with_state.default
});
module.exports = __toCommonJS(index_exports);
__reExport(index_exports, require("./utils/create-higher-order-component/index.cjs"), module.exports);
__reExport(index_exports, require("./utils/debounce/index.cjs"), module.exports);
__reExport(index_exports, require("./utils/throttle/index.cjs"), module.exports);
__reExport(index_exports, require("./utils/observable-map/index.cjs"), module.exports);
var import_private_apis = require("./private-apis.cjs");
var import_compose = __toESM(require("./higher-order/compose.cjs"));
var import_pipe = __toESM(require("./higher-order/pipe.cjs"));
var import_if_condition = __toESM(require("./higher-order/if-condition/index.cjs"));
var import_pure = __toESM(require("./higher-order/pure/index.cjs"));
var import_with_global_events = __toESM(require("./higher-order/with-global-events/index.cjs"));
var import_with_instance_id = __toESM(require("./higher-order/with-instance-id/index.cjs"));
var import_with_safe_timeout = __toESM(require("./higher-order/with-safe-timeout/index.cjs"));
var import_with_state = __toESM(require("./higher-order/with-state/index.cjs"));
var import_use_constrained_tabbing = __toESM(require("./hooks/use-constrained-tabbing/index.cjs"));
var import_use_copy_on_click = __toESM(require("./hooks/use-copy-on-click/index.cjs"));
var import_use_copy_to_clipboard = __toESM(require("./hooks/use-copy-to-clipboard/index.cjs"));
var import_use_dialog = __toESM(require("./hooks/use-dialog/index.cjs"));
var import_use_disabled = __toESM(require("./hooks/use-disabled/index.cjs"));
var import_use_event = __toESM(require("./hooks/use-event/index.cjs"));
var import_use_dragging = __toESM(require("./hooks/use-dragging/index.cjs"));
var import_use_focus_on_mount = require("./hooks/use-focus-on-mount/index.cjs");
var import_use_focus_outside = __toESM(require("./hooks/use-focus-outside/index.cjs"));
var import_use_focus_return = __toESM(require("./hooks/use-focus-return/index.cjs"));
var import_use_instance_id = __toESM(require("./hooks/use-instance-id/index.cjs"));
var import_use_isomorphic_layout_effect = __toESM(require("./hooks/use-isomorphic-layout-effect/index.cjs"));
var import_use_keyboard_shortcut = __toESM(require("./hooks/use-keyboard-shortcut/index.cjs"));
var import_use_media_query = __toESM(require("./hooks/use-media-query/index.cjs"));
var import_use_previous = __toESM(require("./hooks/use-previous/index.cjs"));
var import_use_reduced_motion = __toESM(require("./hooks/use-reduced-motion/index.cjs"));
var import_use_state_with_history = __toESM(require("./hooks/use-state-with-history/index.cjs"));
var import_use_viewport_match = __toESM(require("./hooks/use-viewport-match/index.cjs"));
var import_use_resize_observer = __toESM(require("./hooks/use-resize-observer/index.cjs"));
var import_use_async_list = __toESM(require("./hooks/use-async-list/index.cjs"));
var import_use_warn_on_change = __toESM(require("./hooks/use-warn-on-change/index.cjs"));
var import_use_debounce = __toESM(require("./hooks/use-debounce/index.cjs"));
var import_use_debounced_input = __toESM(require("./hooks/use-debounced-input/index.cjs"));
var import_use_throttle = __toESM(require("./hooks/use-throttle/index.cjs"));
var import_use_merge_refs = __toESM(require("./hooks/use-merge-refs/index.cjs"));
var import_use_ref_effect = __toESM(require("./hooks/use-ref-effect/index.cjs"));
var import_use_drop_zone = __toESM(require("./hooks/use-drop-zone/index.cjs"));
var import_use_focusable_iframe = __toESM(require("./hooks/use-focusable-iframe/index.cjs"));
var import_use_fixed_window_list = __toESM(require("./hooks/use-fixed-window-list/index.cjs"));
var import_use_observable_value = __toESM(require("./hooks/use-observable-value/index.cjs"));
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  __experimentalUseDialog,
  __experimentalUseDragging,
  __experimentalUseDropZone,
  __experimentalUseFixedWindowList,
  __experimentalUseFocusOutside,
  compose,
  ifCondition,
  pipe,
  privateApis,
  pure,
  useAsyncList,
  useConstrainedTabbing,
  useCopyOnClick,
  useCopyToClipboard,
  useDebounce,
  useDebouncedInput,
  useDisabled,
  useEvent,
  useFocusOnMount,
  useFocusReturn,
  useFocusableIframe,
  useInstanceId,
  useIsomorphicLayoutEffect,
  useKeyboardShortcut,
  useMediaQuery,
  useMergeRefs,
  useObservableValue,
  usePrevious,
  useReducedMotion,
  useRefEffect,
  useResizeObserver,
  useStateWithHistory,
  useThrottle,
  useViewportMatch,
  useWarnOnChange,
  withGlobalEvents,
  withInstanceId,
  withSafeTimeout,
  withState,
  ...require("./utils/create-higher-order-component/index.cjs"),
  ...require("./utils/debounce/index.cjs"),
  ...require("./utils/throttle/index.cjs"),
  ...require("./utils/observable-map/index.cjs")
});
