// @wordpress/theme/build/theme-provider.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/theme/src/theme-provider.tsx
var theme_provider_exports = {};
__export(theme_provider_exports, {
  ThemeProvider: () => ThemeProvider
});
module.exports = __toCommonJS(theme_provider_exports);
var import_element = require("@wordpress/element");
var import_compose = require("@wordpress/compose");
var import_context = require("./context.cjs");
var import_use_theme_provider_styles = require("./use-theme-provider-styles.cjs");

// packages/style-runtime/src/index.ts
var STYLE_HASH_ATTRIBUTE = "data-wp-hash";
function getRuntime() {
  const globalScope = globalThis;
  if (globalScope.__wpStyleRuntime) {
    return globalScope.__wpStyleRuntime;
  }
  globalScope.__wpStyleRuntime = {
    documents: /* @__PURE__ */ new Map(),
    styles: /* @__PURE__ */ new Map(),
    injectedStyles: /* @__PURE__ */ new WeakMap()
  };
  if (typeof document !== "undefined") {
    registerDocument(document);
  }
  return globalScope.__wpStyleRuntime;
}
function documentContainsStyleHash(targetDocument, hash) {
  if (!targetDocument.head) {
    return false;
  }
  for (const style of targetDocument.head.querySelectorAll(
    `style[${STYLE_HASH_ATTRIBUTE}]`
  )) {
    if (style.getAttribute(STYLE_HASH_ATTRIBUTE) === hash) {
      return true;
    }
  }
  return false;
}
function injectStyle(targetDocument, hash, css) {
  if (!targetDocument.head) {
    return;
  }
  const runtime = getRuntime();
  let injectedStyles = runtime.injectedStyles.get(targetDocument);
  if (!injectedStyles) {
    injectedStyles = /* @__PURE__ */ new Set();
    runtime.injectedStyles.set(targetDocument, injectedStyles);
  }
  if (injectedStyles.has(hash)) {
    return;
  }
  if (documentContainsStyleHash(targetDocument, hash)) {
    injectedStyles.add(hash);
    return;
  }
  const style = targetDocument.createElement("style");
  style.setAttribute(STYLE_HASH_ATTRIBUTE, hash);
  style.appendChild(targetDocument.createTextNode(css));
  targetDocument.head.appendChild(style);
  injectedStyles.add(hash);
}
function registerDocument(targetDocument) {
  const runtime = getRuntime();
  runtime.documents.set(
    targetDocument,
    (runtime.documents.get(targetDocument) ?? 0) + 1
  );
  for (const [hash, css] of runtime.styles) {
    injectStyle(targetDocument, hash, css);
  }
  return () => {
    const count = runtime.documents.get(targetDocument);
    if (count === void 0) {
      return;
    }
    if (count <= 1) {
      runtime.documents.delete(targetDocument);
      return;
    }
    runtime.documents.set(targetDocument, count - 1);
  };
}
function registerStyle(hash, css) {
  const runtime = getRuntime();
  runtime.styles.set(hash, css);
  for (const targetDocument of runtime.documents.keys()) {
    injectStyle(targetDocument, hash, css);
  }
}

// packages/theme/src/style.module.css
if (typeof process === "undefined" || process.env.NODE_ENV !== "test") {
  registerStyle("f4e6e06c6a", ".dba930ea7a9438fd__root{display:contents}");
}
var style_default = { "root": "dba930ea7a9438fd__root" };

// packages/theme/src/theme-provider.tsx
var import_jsx_runtime = require("react/jsx-runtime");
var rootProviderCountByDocument = /* @__PURE__ */ new WeakMap();
var ThemeProvider = ({
  children,
  color = {},
  cursor,
  cornerRadius,
  isRoot = false
}) => {
  const { themeProviderStyles, resolvedSettings } = (0, import_use_theme_provider_styles.useThemeProviderStyles)({
    color,
    cursor,
    cornerRadius
  });
  const cornerRadiusPreset = resolvedSettings.cornerRadius ?? "subtle";
  const contextValue = (0, import_element.useMemo)(
    () => ({
      resolvedSettings
    }),
    [resolvedSettings]
  );
  const wrapperRef = (0, import_element.useRef)(null);
  (0, import_compose.useIsomorphicLayoutEffect)(() => {
    if (!isRoot) {
      return;
    }
    const doc = wrapperRef.current?.ownerDocument;
    if (!doc) {
      return;
    }
    const root = doc.documentElement;
    if (process.env.NODE_ENV !== "production") {
      const active = rootProviderCountByDocument.get(doc) ?? 0;
      if (active > 0) {
        console.warn(
          "ThemeProvider: More than one root provider (`isRoot`) is mounted on the same document. Their forwarded document-level styles conflict, and unmounting one can reset the others. Render a single root provider per document."
        );
      }
      rootProviderCountByDocument.set(doc, active + 1);
    }
    const previous = /* @__PURE__ */ new Map();
    const applied = [];
    for (const [rawKey, rawValue] of Object.entries(
      themeProviderStyles
    )) {
      if (!rawKey.startsWith("--") || rawValue === null || rawValue === void 0) {
        continue;
      }
      previous.set(rawKey, root.style.getPropertyValue(rawKey));
      root.style.setProperty(rawKey, String(rawValue));
      applied.push(rawKey);
    }
    return () => {
      if (process.env.NODE_ENV !== "production") {
        const active = rootProviderCountByDocument.get(doc) ?? 1;
        if (active <= 1) {
          rootProviderCountByDocument.delete(doc);
        } else {
          rootProviderCountByDocument.set(doc, active - 1);
        }
      }
      for (const key of applied) {
        const prev = previous.get(key);
        if (prev) {
          root.style.setProperty(key, prev);
        } else {
          root.style.removeProperty(key);
        }
      }
    };
  }, [isRoot, themeProviderStyles]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      ref: wrapperRef,
      "data-wpds-root-provider": isRoot || void 0,
      "data-wpds-corner-radius": cornerRadiusPreset,
      className: style_default.root,
      style: themeProviderStyles,
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_context.ThemeContext.Provider, { value: contextValue, children })
    }
  );
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ThemeProvider
});
