// @wordpress/compose/build/higher-order/with-state/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/compose/src/higher-order/with-state/index.js
var with_state_exports = {};
__export(with_state_exports, {
  default: () => withState
});
module.exports = __toCommonJS(with_state_exports);
var import_element = require("@wordpress/element");
var import_deprecated = __toESM(require("@wordpress/deprecated"));
var import_create_higher_order_component = require("../../utils/create-higher-order-component/index.cjs");
var import_jsx_runtime = require("react/jsx-runtime");
function withState(initialState = {}) {
  (0, import_deprecated.default)("wp.compose.withState", {
    since: "5.8",
    alternative: "wp.element.useState"
  });
  return (0, import_create_higher_order_component.createHigherOrderComponent)((OriginalComponent) => {
    return class WrappedComponent extends import_element.Component {
      constructor(props) {
        super(props);
        this.setState = this.setState.bind(this);
        this.state = initialState;
      }
      render() {
        return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          OriginalComponent,
          {
            ...this.props,
            ...this.state,
            setState: this.setState
          }
        );
      }
    };
  }, "withState");
}
